File: python/bsearch.py
#!/usr/bin/env python3
import random

random.seed(42)

# Binary Search
def bsearch(A: list[int], target: int) -> int:

	size = len(A)
	low = 0
	high = size - 1

	while(low <= high):
		mid = (low + high) // 2
		if(target == A[mid]):
			print("Element found at index", mid)
			return mid
		else:
			if(target > A[mid]):
				low = mid + 1
			else:
				high = mid - 1
	print("Element not found.")


def main():
	N = 100

	A = sorted([random.randint(1, N + N) for _ in range(N)])
	target = A[-1] # last element in list A (worst case)

	bsearch(A, target)


"""
Time Complexity:

	best case: O(1)
	worst case: O(log N)
	average case: O(log N)

"""

if __name__ == "__main__":
	main()

File: c/bsearch.c
#include <stdio.h>

// Binary Search

int main(){
	int N[7] = {1, 3, 5, 7, 13, 15, 18};
	int target = 3;

	int len = sizeof(N) / sizeof(N[0]);

	int low = 0;
	int high = len - 1;

	while(low <= high){
		int mid = (low + high) / 2;

		if(target == N[mid]){
			printf("Element Found at %d \n", mid);
			return 0;
		}
		else if(target > N[mid]){
			low = mid + 1;
		}
		else{
			high = mid - 1;
		}
	}
	printf("Element not found. \n");
	return 0;
  
}

/*

Time Complexity :

	-> Best Case : O(1)
	-> Worst Case : O(logn) 

*/